### PR TITLE
impl(common): abstractions for asynchronous streaming read RPCs

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #8937 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8935)

### CODE REVIEW

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  std::unique_ptr<grpc::ClientAsyncReaderInterface<Response>> stream_;
+};
+
+template <typename Response>
+using PrepareAsyncReadRpc = absl::FunctionRef<
+    std::unique_ptr<grpc::ClientAsyncReaderInterface<Response>>(
+        grpc::ClientContext*, grpc::CompletionQueue*)>;
+
+/**
+ * Make an asynchronous streaming read/write RPC using `CompletionQueue`.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed.

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * @note In the past we would have made this a member function of the
+ *     `CompletionQueue` class. We want to avoid this as (a) we are not certain
+ *     this is the long term API we want to expose, (b) once in the public
+ *     `CompletionQueue` class it is hard to remove member functions.  Placing
+ *     the API in the `internal::` namespace give us more flexibility for the
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      : cq_(std::move(cq)),
+        context_(std::move(context)),
+        stream_(std::move(stream)) {}
+
+  void Cancel() override { context_->TryCancel(); }
+
+  future<bool> Start() override {
+    struct OnStart : public AsyncGrpcOperation {
+      promise<bool> p;
+      Options options_ = CurrentOptions();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed.

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    CompletionQueue const& cq, std::unique_ptr<grpc::ClientContext> context,
+    PrepareAsyncReadRpc<Response> async_call) {
+  auto cq_impl = GetCompletionQueueImpl(cq);
+  auto stream = async_call(context.get(), &cq_impl->cq());
+  return absl::make_unique<AsyncStreamingReadRpcImpl<Response>>(
+      std::move(cq_impl), std::move(context), std::move(stream));
+}
+
+/**
+ * A streaming read-write RPC returning a fixed error.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed.

File: google/cloud/internal/async_streaming_read_rpc.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+ */
+template <typename Response>
+class AsyncStreamingReadRpc {
+ public:
+  virtual ~AsyncStreamingReadRpc() = default;
+
+  /**
+   * Sends a best-effort request to cancel the RPC.
+   *
+   * The application should still wait for the current operation(s) (any pending
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc.h]: Sure, fixed.

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 6]
```diff
@@ -36,7 +36,7 @@ namespace internal {
  *
  * A wrapper for gRPC's asynchronous streaming read APIs, which can be
  * combined with `google::cloud::CompletionQueue` and `google::cloud::future<>`
- * to provide easier-to-use abstractions.
+ * to provide easier-to-use abstractions than the objeects returned by gRPC.
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_impl.h:39]: s/objeects/objects/

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 7]
```diff
@@ -139,13 +139,13 @@ std::unique_ptr<AsyncStreamingReadRpc<Response>> MakeStreamingReadRpc(
...
 /**
- * A streaming read-write RPC returning a fixed error.
+ * An asynchronous streaming read RPC returning a fixed error.
  *
  * This is used when the library cannot even start the streaming RPC, for
  * example, because setting up the credentials for the call failed.  One could
- * return `StatusOr<std::unique_ptr<StreamingWriteRpc<A, B>>` in such cases, but
- * the receiving code must deal with streams that fail anyway. It seems more
- * elegant to represent the error as part of the stream.
+ * return `StatusOr<std::unique_ptr<StreamingReadRpc<Response>>` in such cases.
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_impl.h:146]: s/StreamingReadRpc/AsyncStreamingReadRpc/

**Review by dbolduc — APPROVED** [review]

File: google/cloud/internal/async_streaming_read_rpc_impl.h (hunk 6 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed.

File: google/cloud/internal/async_streaming_read_rpc_impl.h (hunk 7 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_impl.h]: Fixed.

**Pull request merged**